### PR TITLE
New Resource: aws_ec2_fleet

### LINKED ISSUE #4581: EC2 Fleet Resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

EC2 Fleet gives the ability to create fleets that are composed of a combination of EC2 On-Demand, Reserved, and Spot Instances with a single API call. 

https://aws.amazon.com/blogs/aws/ec2-fleet-manage-thousands-of-on-demand-and-spot-instances-with-one-request/

### New or Affected Resource(s)

* aws_ec2_fleet

### Potential Terraform Configuration

```hcl
resource "aws_launch_template" "foo" {
...
}

resource "aws_ec2_fleet" "foo" {
  request_type = "maintain"

  launch_template = {
    id = "${aws_launch_template.foo.id}"
    version = "$$Latest"
  }

  launch_template_overrides {
    instance_type = "m4.16xlarge"
    weighted_capacity = 64
  }

  launch_template_overrides {
    instance_type = "m5.24xlarge"
    weighted_capacity = 96
  }

  total_target_capacity = 2880
  on_demand_target_capacity = 960
  spot_target_capacity = 1920
  default_target_capacity_type = "spot"
}
```

### References

* https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-fleet.html

### PR DESCRIPTION
**Pull request by bflad**
Closes #4581 

Changes proposed in this pull request:

* New Resource: `aws_ec2_fleet`

Output from acceptance testing:

```
Tests passed: 25, ignored: 1
--- PASS: TestAccAWSEc2Fleet_basic (26.31s)
--- PASS: TestAccAWSEc2Fleet_disappears (26.95s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_InstanceType (49.06s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_LaunchTemplateSpecification_LaunchTemplateId (55.89s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_LaunchTemplateSpecification_LaunchTemplateName (56.60s)
--- PASS: TestAccAWSEc2Fleet_ExcessCapacityTerminationPolicy (58.72s)
--- PASS: TestAccAWSEc2Fleet_Tags (59.63s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_AllocationStrategy (62.09s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_LaunchTemplateSpecification_Version (61.98s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType_OnDemand (35.94s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType (62.78s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_WeightedCapacity (62.83s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_InstancePoolsToUseCount (63.36s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_InstanceInterruptionBehavior (63.30s)
--- PASS: TestAccAWSEc2Fleet_ReplaceUnhealthyInstances (65.50s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_Priority (66.12s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_Priority_Multiple (66.07s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType_Spot (40.08s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_SubnetId (67.60s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_WeightedCapacity_Multiple (67.61s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfigs_Overrides_AvailabilityZone (76.65s)
--- PASS: TestAccAWSEc2Fleet_OnDemandOptions_AllocationStrategy (79.66s)
--- PASS: TestAccAWSEc2Fleet_Type (25.54s)
--- PASS: TestAccAWSEc2Fleet_TerminateInstancesWithExpiration (47.88s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_TotalTargetCapacity (423.74s)
```

### CODE REVIEW

**Comment by zytek** [comment]
Does this supersede ec2_spot_fleet ?  Seems so.

File: aws/resource_aws_ec2_fleet.go [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+			"excess_capacity_termination_policy": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  ec2.FleetExcessCapacityTerminationPolicyTermination,
+				ValidateFunc: validation.StringInSlice([]string{
+					ec2.FleetExcessCapacityTerminationPolicyNoTermination,
+					ec2.FleetExcessCapacityTerminationPolicyTermination,
+				}, false),
+			},
+			"launch_template_configs": {
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: should be singular

File: aws/resource_aws_ec2_fleet.go [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+									},
+									"version": {
+										Type:     schema.TypeString,
+										Required: true,
+										ForceNew: true,
+									},
+								},
+							},
+						},
+						"overrides": {
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: should be singular

File: aws/resource_aws_ec2_fleet.go [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+										Optional: true,
+										ForceNew: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"on_demand_options": {
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: should be singular

File: aws/resource_aws_ec2_fleet.go [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+						},
+					},
+				},
+			},
+			"replace_unhealthy_instances": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+			"spot_options": {
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: should be singular

File: aws/resource_aws_ec2_fleet.go [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+		},
+	}
+}
+
+func resourceAwsEc2FleetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	input := &ec2.CreateFleetInput{
+		ExcessCapacityTerminationPolicy:  aws.String(d.Get("excess_capacity_termination_policy").(string)),
+		LaunchTemplateConfigs:            expandEc2FleetLaunchTemplateConfigRequests(d.Get("launch_template_configs").([]interface{})),
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: I feel like these casts are panic potential, but won't hold up a merge over it

File: aws/resource_aws_ec2_fleet.go [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+			return nil, ec2.FleetStateCodeDeleted, nil
+		}
+
+		var fleet *ec2.FleetData
+		for _, fleetData := range output.Fleets {
+			if fleetData == nil {
+				continue
+			}
+			if aws.StringValue(fleetData.FleetId) == fleetID {
+				fleet = fleetData
```
Comment (paultyng) [inline aws/resource_aws_ec2_fleet.go]: should this `break` here?

File: aws/resource_aws_ec2_fleet.go (hunk 4 above)
Comment (paultyng) [reply aws/resource_aws_ec2_fleet.go]: nevermind, max 1

File: aws/resource_aws_ec2_fleet.go (hunk 1 above)
Comment (paultyng) [reply aws/resource_aws_ec2_fleet.go]: nevermind, max 1

File: aws/resource_aws_ec2_fleet.go (hunk 5 above)
Comment (bflad) [reply aws/resource_aws_ec2_fleet.go]: Calling `d.Get()` against a `schema.TypeList` will always return a `[]interface{}` with 0 or more elements 😄

File: aws/resource_aws_ec2_fleet.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_ec2_fleet.go]: Will update code, testing, and documentation. 👍

File: aws/resource_aws_ec2_fleet.go (hunk 6 above)
Comment (bflad) [reply aws/resource_aws_ec2_fleet.go]: Argh, yes it should!

**Comment by bflad** [comment]
@paultyng feedback hopefully addressed! 😄 

Output from acceptance testing:

```
--- PASS: TestAccAWSEc2Fleet_basic (35.71s)
--- PASS: TestAccAWSEc2Fleet_disappears (32.21s)
--- PASS: TestAccAWSEc2Fleet_ExcessCapacityTerminationPolicy (94.75s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_LaunchTemplateSpecification_LaunchTemplateId (72.23s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_LaunchTemplateSpecification_LaunchTemplateName (65.05s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_LaunchTemplateSpecification_Version (66.15s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_AvailabilityZone (64.70s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_InstanceType (70.11s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_Priority (62.72s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_Priority_Multiple (62.67s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_SubnetId (84.59s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_WeightedCapacity (66.18s)
--- PASS: TestAccAWSEc2Fleet_LaunchTemplateConfig_Override_WeightedCapacity_Multiple (66.18s)
--- PASS: TestAccAWSEc2Fleet_OnDemandOptions_AllocationStrategy (65.99s)
--- PASS: TestAccAWSEc2Fleet_ReplaceUnhealthyInstances (67.76s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_AllocationStrategy (65.17s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_InstanceInterruptionBehavior (64.85s)
--- PASS: TestAccAWSEc2Fleet_SpotOptions_InstancePoolsToUseCount (62.53s)
--- PASS: TestAccAWSEc2Fleet_Tags (65.32s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType (64.69s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType_OnDemand (32.68s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_DefaultTargetCapacityType_Spot (34.97s)
--- PASS: TestAccAWSEc2Fleet_TargetCapacitySpecification_TotalTargetCapacity (468.10s)
--- PASS: TestAccAWSEc2Fleet_TerminateInstancesWithExpiration (63.03s)
--- PASS: TestAccAWSEc2Fleet_Type (32.81s)
```

File: aws/resource_aws_ec2_fleet.go (hunk 1 above)
Comment (bflad) [reply aws/resource_aws_ec2_fleet.go]: You were correct that this one should actually be singular, even with max 1, updated. 👍

**Pull request merged**